core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+        Objects.requireNonNull(request.shardId(), "missing shardId");
+        this.executeOnReplicas = executeOnReplicas;
+        this.replicasProxy = replicasProxy;
+        this.primary = primary;
+        this.finalResponseListener = listener;
+        this.logger = logger;
+        this.request = request;
+        this.clusterStateSupplier = clusterStateSupplier;
+        this.opType = opType;
+    }
+
+
+    // nocommit: deal with exceptions, tricky part is exception during replication logic where operation has already been performed locally
+    public void performOnPrimary() throws Exception {
+
+        final String writeConsistencyFailure = checkWriteConsistency ? checkWriteConsistency() : null;
+        if (writeConsistencyFailure != null) {
+            finishAsFailed(new UnavailableShardsException(
+                    request.shardId(), "{} Timeout: [{}], request: [{}]", writeConsistencyFailure, request.timeout(), request));
+            return;
+        }
+
+        pendingShards.incrementAndGet(); // increase by 1 until we finish all primary coordination
+        Tuple<Response, ReplicaRequest> primaryResponse = primary.performOperation(request);
+        finalResponse = primaryResponse.v1();
+        ReplicaRequest replicaRequest = primaryResponse.v2();
+        replicaRequest.primaryTerm(primary.term());
+        if (logger.isTraceEnabled()) {
+            logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
+        }
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        // If the index gets deleted after primary operation, we skip replication
+        List<ShardRouting> shards = getShards(request.shardId(), clusterStateSupplier.get());
+        final String localNodeId = primary.routingEntry().currentNodeId();
+        for (final ShardRouting shard : shards) {
+            if (executeOnReplicas == false || shard.unassigned()) {
+                if (shard.primary() == false) {
+                    totalShards.incrementAndGet();
+                }
+                continue;
+            }
+
+            if (shard.currentNodeId().equals(localNodeId) == false) {
+                performOnReplica(shard, replicaRequest);
+            }
+
+            if (shard.relocating() && shard.relocatingNodeId().equals(localNodeId) == false) {
+                performOnReplica(shard.buildTargetRelocatingShard(), replicaRequest);
+            }
+        }
+
+        // mark primary as successful, decrement pending and finish (if there are no replicas, or those are done)
+        totalShards.incrementAndGet();
```

Comment:
I think we should follow the pattern for the replicas here, i.e. `totalShards.incrementAndGet();` should be put directly before `pendingShards.incrementAndGet(); ` in this method.